File: src/domain/article.js
import { Locator } from "./locator";
import { Map } from "immutable";

export class Article {
  /**
   * data = Map({ namespace, id, title, text})
   * permissions \in { "none", "readonly", "full" }
   */
  constructor(data, permissions) {
    this.data = ArticleSerializationService.cleanData(data);
    this.permissions = permissions;
  }

  getData() {
    this.ensureIsReadable();
    return this.data;
  }

  getTitleBasedLocator() {
    this.ensureIsReadable();
    return new Locator(this.data.get("namespace"), this.data.get("title"));
  }

  isReadOnly() {
    return this.permissions === "readonly";
  }

  isReadable() {
    return this.permissions === "readonly" || this.permissions === "full";
  }

  isReadableAndWritable() {
    return this.data.permissions === "full";
  }

  ensureIsReadable() {
    if (!this.isReadable()) {
      throw new ForbiddenOperationException();
    }
  }
}

export class ForbiddenOperationException extends Error {}

export class ArticleSerializationService {
  static serialize(article) {
    return JSON.stringify({
      data: ArticleSerializationService.serializeData(article.getData()),
      permissions: article.permissions,
    });
  }

  static deserialize(string) {
    const parsed = JSON.parse(string);
    return new Article(
      ArticleSerializationService.deserializeData(parsed.data),
      parsed.permissions
    );
  }

  static serializeData(data) {
    return JSON.stringify(ArticleSerializationService.cleanData(data));
  }

  static deserializeData(string) {
    return ArticleSerializationService.cleanData(Map(JSON.parse(string)));
  }

  static cleanData(data) {
    return Map({
      namespace: data.get("namespace"),
      id: data.get("id"),
      title: data.get("title"),
      text: data.get("text"),
    });
  }
}

File: server/app/domain/article.py
from .locator import Locator
import json

class Article:
    def __init__(self, data, permissions):
        self.data = ArticleSerializationService.cleanData(data)
        self.permissions = permissions

    def getData(self):
        self.ensureIsReadable()
        return self.data

    def getTitleBasedLocator(self):
        self.ensureIsReadable()
        return Locator(self.data["namespace"], self.data["title"])

    def isReadOnly(self):
        return self.permissions == "readonly"

    def isReadable(self):
        return self.permissions in ["full", "readonly"]

    def isReadableAndWritable(self):
        return self.permissions == "full"

    def ensureIsReadable(self):
        if not self.isReadable():
            raise ForbiddenOperationException()

class ForbiddenOperationException(Exception):
    pass

class ArticleSerializationService:

    @staticmethod
    def serialize(article):
        data = article.getData()
        return json.dumps({
            "data": ArticleSerializationService.serializeData(data),
            "permissions": article.permissions,
        })

    @staticmethod
    def deserialize(string):
        parsed = json.loads(string)
        return Article(ArticleSerializationService.deserializeData(parsed["data"]), parsed["permissions"])

    @staticmethod
    def serializeData(data):
        return json.dumps(ArticleSerializationService.cleanData(data))

    @staticmethod
    def deserializeData(string):
        return ArticleSerializationService.cleanData(json.loads(string))

    @staticmethod
    def cleanData(data):
        return {
            "namespace": data["namespace"],
            "id": data["id"],
            "title": data["title"],
            "text": data["text"],
        }
